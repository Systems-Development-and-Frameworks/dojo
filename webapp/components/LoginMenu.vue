<template>
  <div>
    <nuxt-link v-if="!isAuthenticated" :to="{name: 'login', params: {returnPath: '/'}}">
      <button id="login-button">Login</button>
    </nuxt-link>
    <template v-else>
      <button @click="logout" id="logout-button">Logout</button>
      <span>userId: {{ userId }}</span>
    </template>
  </div>
</template>

<script>
import { mapGetters, mapMutations } from 'vuex'

export default {
  computed: {
    ...mapGetters({
      isAuthenticated: 'auth/isAuthenticated',
      userId: 'auth/userId'
    })
  },
  methods: {
    async logout () {
      this.unsetToken()
      await this.$apolloHelpers.onLogout()
      await this.$nuxt.refresh()
    },
    ...mapMutations({
      unsetToken: 'auth/unsetToken'
    })
  }
}
</script>
