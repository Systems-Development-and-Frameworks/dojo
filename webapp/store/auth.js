import jwt_decode from "jwt-decode"

export const state = () => ({
  token: null,
  userId: null,
})

export const getters = {
  isAuthenticated (state) {
    return !!state.token
  },
  userId (state) {
    return state.userId
  },
  token (state) {
    return state.token
  }
}

export const mutations = {
  _setToken (state, token) {
    state.token = token
  },
  _setUserId (state, userId) {
    state.userId = userId
  },
  unsetToken (state) {
    state.userId = null
    state.token = null
  }
}

export const actions = {
  async setToken (context, token) {
    const { userId } = jwt_decode(token)
    context.commit('_setToken', token)
    context.commit('_setUserId', userId)
  }
}
