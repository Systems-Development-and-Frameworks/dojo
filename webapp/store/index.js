import cookie from 'cookie'

export const actions = {
  async nuxtServerInit (store, context) {
    const { req } = context.ssrContext
    if (!req || typeof req.headers.cookie != 'string') return // static site generation
    const parsedCookies = cookie.parse(req.headers.cookie)
    const token = parsedCookies['apollo-token']
    if (!token) return
    req.headers.authorization = token
    return await store.dispatch('auth/setToken', token)
  },
}
