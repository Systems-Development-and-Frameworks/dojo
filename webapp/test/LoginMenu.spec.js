import '@testing-library/jest-dom'
import 'regenerator-runtime'

import { createLocalVue, shallowMount } from '@vue/test-utils'

import Vuex from 'vuex'

import LoginMenu from '@/components/LoginMenu'
import { actions, getters, mutations, state } from '@/store/auth'

import jwt_decode from 'jwt-decode'

jest.mock('jwt-decode')

describe('LoginMenu', () => {
  const localVue = createLocalVue()
  localVue.use(Vuex)

  let wrapper, store

  const createLoginMenu = async (appOptions) => {
    store = new Vuex.Store({
      modules: {
        auth: {
          namespaced: true,
          state,
          getters,
          actions,
          mutations
        }
      }
    })

    wrapper = shallowMount(LoginMenu, {
      localVue,
      store,
      stubs: {
        NuxtLink: true,
      },
      ...appOptions
    })
    await wrapper.vm.$nextTick()
  }

  afterEach(() => {
    wrapper.destroy()
    store = null
  })

  describe('for a user that is not logged in', () => {
    it('should show only the login button', async () => {
      await createLoginMenu()
      await wrapper.vm.$nextTick()

      expect(wrapper.find('#login-button').exists()).toBe(true)
      expect(wrapper.find('#logout-button').exists()).toBe(false)
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('for a user that is logged in', () => {
    jwt_decode.mockReturnValue({ userId: '1234' })

    it('should show only the logout button and the user id', async () => {
      await createLoginMenu()
      await wrapper.vm.$nextTick()

      await store.dispatch('auth/setToken', 'someToken')

      expect(wrapper.find('#login-button').exists()).toBe(false)
      expect(wrapper.find('#logout-button').exists()).toBe(true)
      expect(wrapper.html()).toMatchSnapshot()
    })

    describe('triggering the logout button', () => {
      let $nuxt, $apolloHelpers
      beforeEach(() => {
        $nuxt = {
          refresh: jest.fn()
        }
        $apolloHelpers = {
          onLogout: jest.fn()
        }
      })

      async function setTokenThenLogout () {
        await store.dispatch('auth/setToken', 'someToken')
        await wrapper.vm.$nextTick()

        await wrapper.find('#logout-button').trigger('click')
        await wrapper.vm.$nextTick()
      }

      it('should remove the token and userId from the state', async () => {
        await createLoginMenu({
          mocks: {
            $apolloHelpers,
            $nuxt
          },
        })

        await setTokenThenLogout()

        expect(store.getters['auth/isAuthenticated']).toBeFalsy()
        expect(store.getters['auth/userId']).toBe(null)
        expect(store.getters['auth/token']).toBe(null)
      })

      it('should call $apolloHelpers.onLogout', async () => {
        await createLoginMenu({
          mocks: {
            $apolloHelpers,
            $nuxt
          },
        })

        await setTokenThenLogout()

        expect($apolloHelpers.onLogout).toBeCalledTimes(1)
      })

      it('should call $nuxt.refresh', async () => {
        await createLoginMenu({
          mocks: {
            $apolloHelpers,
            $nuxt
          },
        })

        await setTokenThenLogout()

        expect($nuxt.refresh).toBeCalledTimes(1)
      })
    })
  })
})

