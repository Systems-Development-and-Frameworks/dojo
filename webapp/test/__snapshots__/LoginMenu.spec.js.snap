// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginMenu for a user that is logged in should show only the logout button and the user id 1`] = `"<div><button id=\\"logout-button\\">Logout</button> <span>userId: 1234</span></div>"`;

exports[`LoginMenu for a user that is not logged in should show only the login button 1`] = `
"<div>
  <nuxtlink-stub to=\\"[object Object]\\"><button id=\\"login-button\\">Login</button></nuxtlink-stub>
</div>"
`;
